numpy/core/src/multiarray/datetime.c

```diff
@@ -2457,8 +2457,14 @@ convert_pyobject_to_datetime(PyArray_DatetimeMetaData *meta, PyObject *obj,
         Py_DECREF(bytes);
         return 0;
     }
+
     /* Do no conversion on raw integers */
-    else if (PyInt_Check(obj) || PyLong_Check(obj)) {
+    PyObject *type, *value, *traceback;
+    PyErr_Fetch(&type, &value, &traceback);
+    npy_intp intp = PyArray_PyIntAsIntp(obj);
+    PyErr_Restore(type, value, traceback);
+
+    if (intp != -1 || PyInt_Check(obj)) {
```

Comment:
No need for the second half of this.
```
if (!error_converting(intp))
```
Would be more idiomatic, and also correctly handles `-1`.